libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Analytics/SFSDKSalesforceAnalyticsManager.m

```diff
@@ -315,13 +313,16 @@ - (void) publishOnAppBackground {
     if (![self.userAccount.accountIdentity isEqual:[SFUserAccountManager sharedInstance].currentUser.accountIdentity]) {
         return;
     }
-    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
-        __block typeof(self) weakSelf = self;
-        task = [[SFApplicationHelper sharedApplication] beginBackgroundTaskWithName:NSStringFromClass([self class]) expirationHandler:^{
-            [weakSelf cleanupBackgroundTask];
-        }];
-        [self publishAllEvents];
-    });
+    // Avoid re-entrance if task is active
+    if (self.task == UIBackgroundTaskInvalid) {
+        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
```

Comment:
There is a small window for a race condition here.

Reply:
Yeah, I saw that as well but seems much less likely that the current state. We can revise in subsequent PRs but need to CP something in 220. 

I thought maybe we can just convert this to use a serial queue but the publish event is async in itself, so the execution of the block would terminate allowing another block in before we end the task on a successful publish. Same issue with a critical section. Have to unravel the multiple asynchronous behaviors here to have a truly bullet proof approach.

Reply:
Sounds good.